Admin/AdminHelper.php

```diff
@@ -108,82 +91,69 @@ public function appendFormFieldElement(AdminInterface $admin, $subject, $element
-            // append a new instance into the object
-            $this->addNewInstance($form->getData(), $fieldDescription);
-            $objectCount++;
-        }
-
-        $this->addNewInstance($form->getData(), $fieldDescription);
+        // append a new instance into the object
+        $this->addNewInstance($admin, $elementId);
 
+        // return new form with empty row
         $finalForm = $admin->getFormBuilder()->getForm();
         $finalForm->setData($subject);
-
-        // bind the data
         $finalForm->setData($form->getData());
 
-        return array($fieldDescription, $finalForm);
+        return $finalForm;
     }
 
     /**
-     * Add a new instance to the related FieldDescriptionInterface value
+     * Add a new instance
      *
-     * @param object                                              $object
-     * @param \Sonata\AdminBundle\Admin\FieldDescriptionInterface $fieldDescription
+     * @param AdminInterface $admin
+     * @param string         $elementId
      *
-     * @throws \RuntimeException
+     * @throws \Exception
      */
-    public function addNewInstance($object, FieldDescriptionInterface $fieldDescription)
+    public function addNewInstance(AdminInterface $admin, $elementId)
     {
-        $instance = $fieldDescription->getAssociationAdmin()->getNewInstance();
-        $mapping  = $fieldDescription->getAssociationMapping();
+        $entity = $admin->getSubject();
 
-        $method = sprintf('add%s', $this->camelize($mapping['fieldName']));
+        $propertyAccessor = new PropertyAccessor();
 
-        if (!method_exists($object, $method)) {
-            $method = rtrim($method, 's');
+        $collection = $propertyAccessor->getValue($entity, $elementId);
 
-            if (!method_exists($object, $method)) {
-                $method = sprintf('add%s', $this->camelize(Inflector::singularize($mapping['fieldName'])));
+        if ($collection instanceof \Doctrine\ORM\ArrayCollection) {
+            $entityClassName = $this->entityClassNameFinder($admin, explode('.', preg_replace('#\[\d*?\]#', '', $elementId)));
+        } elseif ($collection instanceof \Doctrine\ORM\PersistentCollection) {
+            //since doctrine 2.4
+            $entityClassName = $collection->getTypeClass()->getName();
+        } else {
+            return;
+        }
 
-                if (!method_exists($object, $method)) {
-                    throw new \RuntimeException(sprintf('Please add a method %s in the %s class!', $method, ClassUtils::getClass($object)));
-                }
-            }
+        if (!method_exists($collection, 'add')){
+            return;
         }
 
-        $object->$method($instance);
+        $collection->add(new $entityClassName);
+
+        $propertyAccessor->setValue($entity, $elementId, $collection);
+    }
+
+
+    protected function entityClassNameFinder(AdminInterface $admin, $elements)
+    {
+        $element = array_shift($elements);
+
+        if(!is_object($element)){
```

Comment:
this can never be an object -> it will always fail because it must be an array - array_shift(explode('.', $something))